[NgChunMan] iP

NgChunMan opened the pull request:
# Evo

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

#### Evo frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

#### All you need to do is,

1. **Download** it from [here](https://github.com/nus-cs2103-AY2324S1/ip/pull/337).
1. **Double-click** it.
1. **Add** your tasks.
1. Let **Evo** manage your tasks for you 🍁 

#### And it is **FREE**!

#### Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public static void main(String[] args) {
    new Evo(FILE_PATH).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

NgChunMan committed:
Add: Level 0

NgChunMan committed:
Add: Level 1

NgChunMan committed:
Add: Level 2

NgChunMan committed:
Add: Level 3

NgChunMan committed:
Add: Level 4 (Inheritance)

NgChunMan committed:
Add: TextUiTesting

NgChunMan committed:
Add: Level 5 (Error Handling)

NgChunMan committed:
Add: Level 6 (Delete)

NgChunMan committed:
Add level 7 save task functionality

NgChunMan committed:
Add level 8 dates and times

NgChunMan committed:
Add more OOP

NgChunMan committed:
Add divide classes into multiple packages

NgChunMan committed:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552'

* commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552':
  Bump gradle and lib version
  Add Gradle support

NgChunMan committed:
Merge branch add-gradle-support and set up Gradle build and run

NgChunMan committed:
Add JUnit tests to test ToDo and Event class

NgChunMan committed:
Create a JAR file

NgChunMan committed:
Add Javadoc for methods and classes

NgChunMan committed:
Follow the coding standard

NgChunMan committed:
Implement find functionality

NgChunMan committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Follow the coding standard

NgChunMan committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Implement find functionality

# Conflicts:
#	src/main/java/evo/commands/FindCommand.java
#	src/main/java/evo/exceptions/InvalidInputException.java
#	src/main/java/evo/exceptions/MissingTaskToFindException.java
#	src/main/java/evo/exceptions/NoTaskFindException.java

Resolution: keep 'branch-Level-9'

NgChunMan committed:
Add coding standard

KamJiaYue commented on src/main/java/evo/Evo.java:
```diff
@@ -0,0 +1,86 @@
+import evo.parser.Parser;
+import evo.commands.Command;
+import evo.commands.ByeCommand;
+import evo.exceptions.EvoException;
```
I like the way you organize your files in folders. It's very neat to me!

KamJiaYue commented on src/main/java/evo/Evo.java:
```diff
@@ -0,0 +1,86 @@
+import evo.exceptions.EvoException;
+
+/**
+ * Evo is a Personal Assistant Chatbot that helps a person to keep track of various things.
```
Perhaps keeping track of various tasks might sound more specific?

KamJiaYue commented on src/main/java/evo/Evo.java:
```diff
@@ -0,0 +1,86 @@
+
+    /**
+     * The file path where task data is stored.
+     */
```
Good Job on adding Javadocs to class fields!

KamJiaYue commented on src/main/java/evo/Evo.java:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * The TaskList that holds and manages tasks.
+     */
+    private TaskList tasksList;
```
Perhaps taskList sounds more logical?

KamJiaYue commented on src/main/java/evo/Evo.java:
```diff
@@ -0,0 +1,86 @@
+        try {
+            tasksList = new TaskList(storage.loadTasksFromFile());
+        } catch (EvoException e) {
+            ui.showLoadingError();
```
I like how you implemented "Tell, Don't Ask" principle from CS2030S, great job!

KamJiaYue commented on src/main/java/evo/Evo.java:
```diff
@@ -0,0 +1,86 @@
+                c.execute(tasksList, ui, storage);
+                storage.saveTasksInFile(tasksList);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Any reason why you throw RunTimeException rather than EvoException?

KamJiaYue commented on src/main/java/evo/commands/Command.java:
```diff
@@ -0,0 +1,29 @@
+    /**
+     * Constructs a generic Command object.
+     */
+    public Command() {
```
This constructor might be redundant as we cannot instantiate abstract classes

KamJiaYue commented on src/main/java/evo/parser/Parser.java:
```diff
@@ -0,0 +1,207 @@
+        } catch (MissingDurationException missingDurationException) {
+            // Catch the exception when the duration of Event object is missing
+            System.out.println("Duration of this event is incomplete. " +
+                    "Please specify the start date/time and/or end date/time.\n");
+        } catch (MissingTaskToDeleteException missingTaskToDeleteException) {
+            // Catch the exception when user never specifies which task to be deleted
+            System.out.println("Please specify the task to be deleted.\n");
+        } catch (MissingTaskToFindException missingTaskToFindException) {
+            // Catch the exception when user never specifies a keyword to find a task
+            System.out.println("Please specify a keyword to find a task.\n");
+        } catch (MissingTaskToMarkException missingTaskToMarkException) {
+            // Catch the exception when user never specifies which task to be marked
+            System.out.println("Please specify the task to be marked.\n");
+        } catch (MissingTaskToUnmarkException missingTaskToUnmarkException) {
+            // Catch the exception when user never specifies which task to be unmarked
+            System.out.println("Please specify the task to be unmarked.\n");
+        } catch (MissingToDoDescriptionException missingToDoDescriptionException) {
+            // Catch the exception when the description of ToDo task is missing
+            System.out.println("Description of this task is missing. " +
+                    "Please specify the description of this task.\n");
+        } catch (NoTaskDeleteException noTaskDeleteException) {
+            // Catch the exception when user tries to delete task from an empty taskList
+            System.out.println("This task cannot be deleted as there is no task in the list.\n");
+        } catch (NoTaskFindException noTaskFindException) {
+            // Catch the exception when user tries to find task from an empty taskList
+            System.out.println("No task can be found as there is no task in the list.\n");
+        } catch (NoSuchTaskException noSuchTaskException) {
+            // Catch the exception when the task to be marked/unmarked does not exist.
+            System.out.println("The task to be marked/unmarked does not exist.\n");
+        } catch (NoSuchTaskDeleteException noSuchTaskDeleteException) {
+            // Catch the exception when the task to be deleted does not exist.
+            System.out.println("The task to be deleted does not exist.\n");
+        } catch (InvalidOperationException invalidOperationException) {
+            // Catch the exception when the operation is invalid
+            System.out.println("OOPS!!! I'm sorry, but I don't know what that means :-(\n");
```
Good practice for handling exceptions!

NgChunMan committed:
Add Checkstyle and fix all the coding style violations

newway1814 commented on src/main/java/evo/commands/EventCommand.java:
```diff
@@ -0,0 +1,87 @@
+ * When executed, it parses user input to create an event task, adds it to the task list, and displays a
+ * confirmation message to the user.
+ */
+public class EventCommand extends Command {
```
Good Job on javadocs and especially adding javadocs for the fields also!!

newway1814 replied:
Good Job on javadocs and especially adding javadocs for the fields also!!

newway1814 commented on src/main/java/evo/exceptions/MissingDescriptionAndDurationException.java:
```diff
@@ -0,0 +1,17 @@
+        super();
+    }
+
+}
```
I liked how you have taken the almost all the exceptions into account.

newway1814 commented on src/main/java/evo/parser/Parser.java:
```diff
@@ -0,0 +1,241 @@
+                // Adds deadline Command
+                if (actionType.length == 1) {
+                    throw new MissingDescriptionAndDeadlineException();
+                }
```
Maybe Could have switch-case statements also.

newway1814 commented on src/main/java/evo/tasks/Event.java:
```diff
@@ -0,0 +1,110 @@
+     * The output message includes task details such as task type, completion status, description, and date/time.
+     *
+     * @return A formatted string representing the Event task details.
+     * @throws InvalidDateAndTimeInputException If the input duration is not in the expected format or if the date/time
```
Could have formatted this javadoc well.

newway1814 commented on src/main/java/evo/ui/Ui.java:
```diff
@@ -0,0 +1,87 @@
+package evo.ui;
```
I think you have done a great job, your code is very clear and it didn't take me much time to trace through your code👍.

NgChunMan committed:
Add GUI to the chatbot

NgChunMan committed:
Add styles

The user interface is a bit plain.

Add styles so that the application looks more attractive.

NgChunMan committed:
Add assertions to document assumptions

* add assertion statements at the point to check the array
length and the validity of task number

These parts consist of checking the input length of
actionType array is of valid length or not and also
checking the task number is greater than one or not.

NgChunMan committed:
Improve code quality

* Refactor key functions to enhance readability and
maintainability
* Resolve code style inconsistencies
* Conduct thorough code review
* Ensure proper error handling and edge case coverage

These changes improve the overall quality of the codebase
and making it easier to maintain and understand.

NgChunMan committed:
Merge pull request #2 from NgChunMan/branch-A-Assertions

Add assertions to document assumptions

NgChunMan committed:
Merge pull request #1 from NgChunMan/branch-A-CodeQuality

Improve code quality

NgChunMan committed:
Add an extension which is DoAfterTask

This commit introduces the DoAfterCommand class, which
handles the creation of 'Do After' tasks and updates the task
list in storage. The class parses user input, constructs new
DoAfter tasks, and provides user feedback. It also includes a
utility method for string concatenation.

* Implement DoAfterCommand class
* Add parsing of user input for 'Do After' tasks
* Create and added 'Do After' tasks to the task list
* Update storage with the new task data
* Add a utility method for string concatenation

NgChunMan committed:
Add tittle to the chatbot

NgChunMan committed:
Add user guide for the evo chatbot

NgChunMan committed:
Make small changes on the style of the chatbot and the output of find functionality

NgChunMan committed:
Change to bullet point representation for each features in user guide

NgChunMan committed:
Modify the code to solve the file path problem

NgChunMan committed:
Improve code quality by shortening too long methods

NgChunMan committed:
Fix javafx dependencies issue

NgChunMan committed:
Update readme file

NgChunMan committed:
Add image of launching page

NgChunMan committed:
Update README.md
